tests/Unit/EmogrifierTest.php

```diff
@@ -684,6 +684,92 @@ public function emogrifyNotAppliesCssToNonMatchingElements($css, $expectedHtml)
     }
 
+    /**
+     * Data provider for {@see EmogrifierTest::emogrifyAppliesMostSpecificSelectorCssToMatchingElements}
+     *
+     * @return string[][]
+     */
+    public function selectorSpecificityCssDataProvider()
+    {
+        /**
+         * @var string[][] $selectorData Data for constructing the parameter sets, whose keys are a test description,
+         *      and values are an array with the following keys:
+         *      0 - first selector expression
+         *      1 - second selector expression
+         *      2 - operator, '<', '==' or '>', defining whether the first selector should respectively have lower,
+         *          equal or higher specificity than the second
+         *      3 - tag (in {@see EmogrifierTest::COMMON_TEST_HTML}) expected to be matched by both selectors, without
+         *          the closing '>', e.g. '<p class="p=1"'
+         */
+        $selectorData = [
+            'class more specific than types' => ['.p-1', 'html body p', '>', '<p class="p-1"'],
+            'id more specific than class and types' => ['#p4', 'html body .p-4', '>', '<p class="p-4" id="p4"'],
+            'id more specific than 20 classes and types'
+                => ['#p4', 'html body ' . str_repeat('.p-4', 20), '>', '<p class="p-4" id="p4"'],
+            'pseudo-class as specific as class' => ['.p-1', '*:first-child', '==', '<p class="p-1"'],
+            'attribute as specific as class' => ['span[title="bonjour"]', '.p-2 span', '==', '<span title="bonjour"'],
+            ':not alone does not increase specificity' => ['p:not(* + *)', 'p', '==', '<p class="p-1"'],
+            ':not with type more specific than nothing' => ['.p-1:not(h1)', '.p-1', '>', '<p class="p-1"'],
+            ':not with type as specific as type' => ['*:not(h1)', 'p', '==', '<p class="p-1"'],
+            ':not with several types more specific than single type'
+                => ['*:not(html body h1)', 'p', '>', '<p class="p-1"'],
+            ':not with class more specific than types'
+                => ['*:not(.p-2)', 'html body p', '>', '<p class="p-1"'],
+            ':not with class more specific than :not with types'
+                => ['*:not(.p-2)', '*:not(html body h1)', '>', '<p class="p-1"'],
+            'class more specific than :not with types' => ['.p-1', '*:not(html body h1)', '>', '<p class="p-1"'],
+        ];
+
+        /**
+         * @var string[][] $parameterSets Sets of parameters returned by this data provider for
+         *      {@see EmogrifierTest::emogrifyAppliesMostSpecificSelectorCssToMatchingElements}.
+         *      The sprintf placeholders %1$s and %2$s will automatically be replaced with CSS declarations
+         *      like 'color: red;' or 'color: blue;'.
+         */
+        $parameterSets = [];
+        foreach ($selectorData as $name => $data) {
+            $firstRuleBlock = $data[0] . ' { %1$s }';
```

Comment:
Maybe also move the code in the loop into a well-named method.

Reply:
Refactored loop body into new private method `getSelectorSpecificityCssDataSets`